fast_llm/models/gpt/trainer.py

```diff
@@ -2,17 +2,35 @@
 import typing
 
 from fast_llm.data.data.gpt.data import GPTData
+from fast_llm.engine.base_model.config import Preprocessor
 from fast_llm.engine.distributed.config import PhaseType
 from fast_llm.engine.training.trainer import Trainer
+from fast_llm.layers.language_model.config import LanguageModelKwargs
 from fast_llm.models.gpt.config import GPTTrainerConfig
-from fast_llm.models.gpt.model import GPTModel
+from fast_llm.models.gpt.model import GPTInferenceRunner
 
 logger = logging.getLogger(__name__)
 
 
+class GPTReferenceModelPreprocessor(Preprocessor):
+    def __init__(self, name: str, inference_runner: GPTInferenceRunner):
+        self._name = name
+        self._inference_runner = inference_runner
+
+    def preprocess_meta(self, kwargs: dict[str, typing.Any]) -> None:
+        pass
+
+    def preprocess(self, batch, kwargs: dict[str, typing.Any]) -> None:
+        # TODO: Fix random state/iteration.
+        preprocess_kwargs = kwargs.copy()
+        del preprocess_kwargs[LanguageModelKwargs.labels]
+        self._inference_runner.forward(batch, preprocess_kwargs, iteration=1)
+        # TODO: Improve.
+        kwargs[f"{self._name}_logits"] = preprocess_kwargs["logits"]
```

Comment:
This works well for now, since we'll need the logits downstream.

One small note: because the reference model has a user-defined name, the distillation loss in #214 and the DPO loss in #209 will need to use that name to access the logits. So the user-defined name should also be passed as an input to the loss functions, which complicates configuration.

Reply:
Maybe not, we can probably get away with a hard-coded name or field default on the loss side, just like we enforce the training dataset to be called "training"